tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -56,6 +58,31 @@ void SetLazyCommandQueueMode(bool lazy) {
     DispatchStateCheck(true);
     LAZY_COMMAND_QUEUE_MODE = lazy;
 }
+
+void ValidateBufferRegion(
+    const std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>>& buffer, const BufferRegion region) {
+    const auto [is_partial_region, is_valid_region, is_sharded_buffer] = std::visit(
+        [&](auto&& buf) -> std::tuple<bool, bool, bool> {
+            using T = std::decay_t<decltype(buf)>;
+            if constexpr (std::is_same_v<T, std::reference_wrapper<Buffer>>) {
+                return {
+                    buf.get().is_partial_region(region),
+                    buf.get().is_valid_region(region),
+                    is_sharded(buf.get().buffer_layout())};
+            } else if constexpr (std::is_same_v<T, std::shared_ptr<Buffer>>) {
+                return {buf->is_partial_region(region), buf->is_valid_region(region), is_sharded(buf->buffer_layout())};
+            }
+        },
+        buffer);
```

Comment:
Can we refactor this to extract the buffer object, and then check the regions after?
Ex. from `enqueue_write_buffer`
```
auto& b = std::visit([&](auto&& b) -> Buffer& {
        using type_buf = std::decay_t<decltype(b)>;
        if constexpr (std::is_same_v<type_buf, std::shared_ptr<Buffer>>) {
            return *b;
        } else {
            return b.get();
        }
    }, buffer);
```